yt/chyt/server/query_analyzer.cpp

```diff
@@ -728,6 +744,125 @@ void TQueryAnalyzer::OptimizeQueryProcessingStage()
+        if (IsFloatingPointType(schemaColumn.CastToV1Type()) && !orderByElement->nulls_direction_was_explicitly_specified) {
+            floatKeyColumnsWithoutExplicitNullsDirection.push_back(schemaColumn.Name());
+        }
+    }
+
+    // See the last comment in this function.
+    if (!floatKeyColumnsWithoutExplicitNullsDirection.empty()) {
+        THROW_ERROR_EXCEPTION(
+            "Nulls direction must be specified explicitly for key columns of floating-point "
+            "types when using optimize_read_in_order is enabled; see docs for more details")
+            << TErrorAttribute("problematic_columns", floatKeyColumnsWithoutExplicitNullsDirection);
+    }
+
+    // In YT, NULL values compare less than any other values. In ClickHouse, NULL value placement
+    // depends solely on the placement requested in the query, which defaults to NULLS LAST.
+    // Since we combine both YT and CH sorting in this mode, there is only one acceptable direction
+    // depending on the requested sort order. In cases when the unsupported option turns out to be
+    // the default (for ascending sorts), we tweak the query to avoid additional user turmoil.
+    if (commonDirection != EReadInOrderMode::None) {
+        for (const auto& orderByElementAst : selectQuery->orderBy()->children) {
+            auto orderByElement = orderByElementAst->as<DB::ASTOrderByElement>();
+
+            // The value of nulls_direction is equal to direction for NULLS LAST and to the opposite
+            // of direction for NULLS FIRST.
+            // For ASC (d = 1), we need NULLS FIRST (nd = -d = -1).
+            // For DSC (d = -1), we need NULLS LAST (nd =  d = -1).
+            if (orderByElement->nulls_direction != -1) {
+                THROW_ERROR_EXCEPTION_IF(
+                    orderByElement->nulls_direction_was_explicitly_specified,
+                    "Specified nulls direction is incompatible with enabled optimize_read_in_order mode; "
+                    "please turn it off via query settings, specify another nulls direction or use the default nulls direction");
+                YT_LOG_DEBUG(
+                    "Reversing ORDER BY nulls direction (ColumnName: %v, Direction: %v, NullsDirection: %v -> -1)",
+                    orderByElement->children.front()->getColumnName(),
+                    orderByElement->direction,
+                    orderByElement->nulls_direction);
+                orderByElement->nulls_direction = -1;
+                orderByElement->nulls_direction_was_explicitly_specified = true;
+            }
+        }
+    }
+    // Sadly, with NANs we are just screwed. In YT these values are larger than any others, but in
+    // CHYT they are always placed together with NULLs. These two approaches contradict each other.
+    // To warn about such cases without disabling this optimization for tables with floating-point
+    // key columns completely, we require to specify nulls direction explicitly as a green light.
```

Comment:
I offer to introduce a separate option for `float` behavior too, so we can safely turn `optimize_read_in_order` on by default